proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -1352,6 +1354,544 @@ lemma scheduleUsed_corres:
+  apply (rule_tac Q="\<lambda>s'. is_active_sc' (ksCurSc s') s'" in corres_cross_add_guard)
+   apply (fastforce intro: is_active_sc'_cross simp: state_relation_def)
+  apply (clarsimp simp: handle_overrun_loop_body_def handleOverrunLoopBody_def)
+  apply (rule corres_split'[rotated 2, OF gets_sp getCurSc_sp])
+   apply (corressimp corres: getCurSc_corres)
+  apply (rule corres_split'[rotated 2, OF refill_single_sp refillSingle_sp])
+   apply (corressimp corres: refillSingle_corres)
+   apply (fastforce dest!: valid_objs'_valid_refills'
+                     simp: obj_at_simps valid_refills'_def opt_map_red)
+  apply (rule corres_split'[rotated 2, OF get_sched_context_sp get_sc_sp'])
+   apply (corressimp corres: get_sc_corres)
+  apply (rename_tac sc sc')
+  apply (rule_tac Q="\<lambda>_ s. sc_refills sc \<noteq> []"
+              and Q'="\<lambda>_ s. valid_sched_context' sc' s"
+              and r'=dc
+               in corres_split'[rotated])
+     apply (fastforce dest: refill_hd_relation2)
+    apply (wpsimp simp: update_refill_hd_def
+                    wp: update_sched_context_wp)
+    apply (clarsimp simp: vs_all_heap_simps obj_at_def)
+   apply wpsimp
+   apply (fastforce dest: sc_ko_at_valid_objs_valid_sc')
+  apply (rule corres_if_split; simp?)
+   apply (corressimp corres: updateRefillHd_corres)
+   apply (fastforce intro!: valid_objs'_valid_refills'
+                      simp: refill_map_def sc_relation_def)
+  apply (rule_tac F="1 < scRefillCount sc'" in corres_req)
+   apply (frule_tac scp="scPtr" and P="\<lambda>l. 1 < l" in length_sc_refills_cross)
+      apply (clarsimp simp: state_relation_def)
+     apply simp
+    apply (fastforce intro!: valid_objs'_valid_refills'
+                       simp: refill_map_def sc_relation_def)
+   apply (clarsimp simp: opt_map_red vs_all_heap_simps obj_at'_def projectKOs Suc_lessI)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split[OF refillPopHead_corres])
+      apply (rule scheduleUsed_corres)
+       apply simp
+      apply (clarsimp simp: refill_map_def sc_relation_def)
+     apply wpsimp
+    apply wpsimp
+   apply (clarsimp simp: is_active_sc2_def sc_at_ppred_def obj_at_def is_sc_obj_def active_sc_def
+                         vs_all_heap_simps opt_map_red Suc_lessI)
+  apply (clarsimp simp: obj_at_simps)
+  done
+
+lemma schedule_used_no_fail[wp]:
+  "no_fail (\<lambda>s. (\<exists>sc n. kheap s sc_ptr = Some (Structures_A.SchedContext sc n)))
+           (schedule_used sc_ptr new)"
+  apply (wpsimp simp: schedule_used_defs get_refills_def refill_full_def
+                  wp: update_sched_context_wp)
+  done
+
+(* FIXME RT: keep and move the following no_fail lemmas? *)
+lemma no_fail_skip:
+  "\<lbrakk>no_fail P f; f \<lbrace>P\<rbrace>; \<And>rv. no_fail P (g rv)\<rbrakk> \<Longrightarrow> no_fail P (f >>= (\<lambda>rv. g rv))"
+  by (wpsimp, blast+)
+
+lemmas no_fail_bind'
+  = no_fail_bind[rotated 2, where P=P and Q=P for P, simplified pred_conj_def, simplified]
```

Comment:
I suppose we can keep these but please don’t rotate `no_fail_bind’`. We don’t want such difference between `no_fail_bind` and `no_fail_bind’`, right? Please keep the precondition order the same.

Reply:
Removed the `rotated 2`